From 4a67109576392d490a0186fb973b7b233387e0e8 Mon Sep 17 00:00:00 2001
From: Mingjing Zhang <mjzhang@chromium.org>
Date: Tue, 03 Jun 2025 14:20:15 -0700
Subject: [PATCH] Add lock for scoped access functions in ClientSharedImage

This CL adds a lock in ClientSharedImage::BeginAccess()/EndAccess(),
in order to guard against possible race conditions accessing variables
in those functions.

Bug: 421926725
Change-Id: Ia9a8cb9274ff63baf22a0d11a22ad9678adb8420
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/6617622
Reviewed-by: Vasiliy Telezhnikov <vasilyt@chromium.org>
Commit-Queue: Mingjing Zhang <mjzhang@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1468960}
---

--- a/gpu/command_buffer/client/client_shared_image.cc
+++ b/gpu/command_buffer/client/client_shared_image.cc
@@ -509,6 +509,7 @@
 }
 
 void ClientSharedImage::BeginAccess(bool readonly) {
+  base::AutoLock lock(lock_);
   if (readonly) {
     CHECK(!has_writer_ ||
           usage().Has(SHARED_IMAGE_USAGE_CONCURRENT_READ_WRITE));
@@ -522,6 +523,7 @@
 }
 
 void ClientSharedImage::EndAccess(bool readonly) {
+  base::AutoLock lock(lock_);
   if (readonly) {
     CHECK(num_readers_ > 0);
     num_readers_--;
--- a/gpu/command_buffer/client/client_shared_image.h
+++ b/gpu/command_buffer/client/client_shared_image.h
@@ -384,10 +384,12 @@
   uint32_t texture_target_ = 0;
 
   // The number of active scoped read accesses.
-  unsigned int num_readers_ = 0;
+  unsigned int num_readers_ GUARDED_BY(lock_) = 0;
 
   // Whether there exists an active scoped write access.
-  bool has_writer_ = false;
+  bool has_writer_ GUARDED_BY(lock_) = false;
+
+  base::Lock lock_;
 };
 
 struct GPU_EXPORT ExportedSharedImage {
